### ISSUE TITLE
A11y: Elements lose visible focus when using a screenreader and hitting Enter to activate

### ISSUE THREAD

**Issue by ericenns** [facebook/docusaurus#11314]
### Have you read the Contributing Guidelines on issues?

- [x] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#issues).

### Prerequisites

- [x] I'm using the latest version of Docusaurus.
- [x] I have tried the `npm run clear` or `yarn clear` command.
- [x] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [x] I have tried creating a repro with https://new.docusaurus.io.
- [x] I have read the console error message carefully (if applicable).

### Description

When navigating a docusaurus site via keyboard while using a screen reader (e.g. NVDA), tabbing to a button and hitting enter causes visible focus to be lost. This is due to https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-theme-common/src/hooks/useKeyboardNavigation.ts, which removes `navigation-with-keyboard` class when clicking, this works for keyboard navigation while not using a screen reader but when using a screen reader hitting enter instead activates mousedown events.

Adding the following css appears to fix the issue:
```css
body:not(.navigation-with-keyboard) *:not(input):focus-visible {
  outline: auto !important;
}
```

Which as you can see uses the `:focus-visible` selector which keeps the outline even if the element sees a mousedown event it will maintain focus if the element was already focused by tabbing to it. If a user clicks on a button not focused it will not have a visible focus ring.

### Reproducible demo

The live docusaurus site exhibits this behaviour so it can be used.

### Steps to reproduce

1. On windows activate [NVDA](https://www.nvaccess.org/download/)
2. Open your browser to https://docusaurus.io/docs
3. Tab to the `collapse` button beside `Getting Started` in the sidebar
4. Hit the enter key

### Expected behavior

Visible focus should be retained

### Actual behavior

Visible focus is lost

### Your environment

- Public source code: N/A
- Public site URL: https://docusaurus.io/docs
- Docusaurus version used: 3.8.1
- Environment name and version (e.g. Chrome 89, Node.js 16.4): N/A
- Operating system and version (e.g. Ubuntu 20.04.2 LTS): N/A

### Self-service

- [x] I'd be willing to fix this bug myself.

**Comment by Huzaif-nabi** [comment]
I can reproduce this and would like to submit a pull request with the suggested fix.

**Comment by OfficialMutazAwad** [comment]
New reply to your ticket #11974

Title: Re: [facebook/docusaurus] A11y: Elements lose visible focus when using a screenreader and hitting Enter to activate (Issue #11314)

Message: Hello,we received your message and our support will reach out to you as soon as possible.

**Comment by slorber** [comment]
Can this be reproduced only on Windows with NVDA?

On macOS with VoiceOver I still see the focus ring on the sidebar item collapse buttons.

Is it related to using the screen reader, or can you reproduce in Windows by simply tab/enter without the screen reader being enabled?

Can you tell us which browser you are using and which version?

**Comment by OfficialMutazAwad** [comment]
New reply to your ticket #12174

Title: Re: [facebook/docusaurus] A11y: Elements lose visible focus when using a screenreader and hitting Enter to activate (Issue #11314)

Message: Hello,we received your message and our support will reach out to you as soon as possible.

**Comment by ericenns** [comment]
> Can this be reproduced only on Windows with NVDA?
> 
> On macOS with VoiceOver I still see the focus ring on the sidebar item collapse buttons.
> 
> Is it related to using the screen reader, or can you reproduce in Windows by simply tab/enter without the screen reader being enabled?
> 
> Can you tell us which browser you are using and which version?

This can also be reproduced on Windows with JAWS

The issue only occurs when using the screen reader, as both NVDA and JAWS through the accesibility api trigger a `click` when hitting `Enter` key.

This occurs on all versions of Chrome and Edge, I haven't tested with Firefox.

**Comment by nmggithub** [comment]
## TL;DR:

I believe the `useKeyboardNavigation` hook is poorly designed for the modern web. At *best*, it provides an aesthetic benefit by removing Chrome's own focus ring when users click non-input focused elements that don't navigate away. At worst, it negatively affect downstream styles (see: https://github.com/electron/website/issues/1001) as well as screen readers that cause `mousedown` events to be emitted.

I see two options:

1. Remove the hook entirely and let downstream users decide how they want to style focus rings.
2. Use something like `:focus-visible` (perhaps instead of listening to mouse events).

## Existing behavior

### With VoiceOver

Below is a video of me using VoiceOver around the button discussed in the initial issue comment. As can be seen, both Chrome and VoiceOver itself add focus indicators. The first toggle is achieved by pressing `Enter`. The second toggle is achieved by clicking. The click removes Chrome's focus ring (due to hook's CSS).

https://github.com/user-attachments/assets/4698d7ba-72d4-4d64-98ea-90d04269cbd6

### Without VoiceOver (and removing the hook's styles)

If I remove the hook's styles, navigate to the same button with the keyboard, and click it, there does appear to be two focus indicators (blue and white) for different states:

https://github.com/user-attachments/assets/311a425c-32b8-4f2c-b9dd-276709017b1d

Removing the hook's styles and testing out tabs (one of the original motivating components for these style changes), the look is less ugly than before. To note, in the below video, I clicked right outside the tabs, pressed the `Tab` key to select the first tab, and then click that first tab.

https://github.com/user-attachments/assets/2562a1b4-e8b7-4337-a0c6-e6cb3640b08e

## Looking to the future

I think the `useKeyboardNavigation` implementation needs to be rethought. CSS pseudo-selectors like `:focus-visible` provide a much better API surface with which to handle focus rings. While listening to `mousedown` events may seem like a good indication that the user is now using a mouse, it's not foolproof. [Some screen readers / assistive technologies simulate mouse actions, thus causing events like `mousedown` to be fired even when the user actually isn't using a keyboard.](https://vispero.com/resources/event-handling-in-jaws-and-nvda/) That previously-linked article explicitly calls out how the `Enter` key resulted in a `mousedown` event firing.

The hook appears to have been made before `:focus-visible` was widely supported ([although there does appear to have been discussion about `:focus-visible` in the issue that prompted the creation of what would eventually become the hook](https://github.com/facebook/docusaurus/issues/3150#issuecomment-665823567)). Regardless, we're kind of fighting the browser (and sometimes downstream devs) with these style changes. A more thoughtful approach is warranted.

**Comment by slorber** [comment]
I agree with you @nmggithub, this hook is not ideal and was written at a time where `:focus-visible` was early and afaik not widely supported.

In https://github.com/facebook/docusaurus/pull/3626 I even proposed to use `:focus-visible` but it was too early

Can you confirm that moving to `:focus-within` would fix the current windows-only problem? Do you want to submit a PR?

---


Note: here's the original blog post because the link in code is broken: https://david-gilbertson.medium.com/removing-that-ugly-focus-ring-and-keeping-it-too-6c8727fefcd2

**Comment by OfficialMutazAwad** [comment]
New reply to your ticket #16067

Title: Re: [facebook/docusaurus] A11y: Elements lose visible focus when using a screenreader and hitting Enter to activate (Issue #11314)

Message: Hello,we received your message and our support will reach out to you as soon as possible.

**Comment by nmggithub** [comment]
@slorber Thanks for following up (and confirming that change might be needed)! I'll do some testing myself and see if I can come up with a better approach by using `:focus-visible` (or maybe `:focus-within`, as you mentioned, but that may have been a typo).

**Comment by slorber** [comment]
@ericenns can you please test on our website docusaurus.io if you still encounter the windows screen reader problem above?

**Comment by ericenns** [comment]
@slorber I tested docusuarus.io and I can confirm that visible focus is no longer lost while using either NVDA or JAWS screen readers on Windows with Chrome.

**Comment by slorber** [comment]
awesome, thanks for confirming 👍
